### PR TITLE
Group managers

### PR DESCRIPTION
**Pull request by TheHollidayInn**
[//]: # (Note: See http://habitica.wikia.com/wiki/Using_Habitica_Git#Pull_Request for more info)

[//]: # (Put Issue # or URL here, if applicable. This will automatically close the issue if your PR is merged in)
Fixes put_issue_url_here

### Changes
[//]: # (Describe the changes that were made in detail here. Include pictures if necessary)
This allows users to be assigned as managers and allows those managers to:
 - create, edit, delete group tasks
 - Assign and unassign users
 - Approve approval requests

[//]: # (Put User ID in here - found in Settings -> API)

----
UUID:

### CODE REVIEW

**Comment by Alys** [comment]
- [x] Before this is merged, I'd like to get feedback from a couple of the Socialites about the UI aspects. I can do that by putting it on a local install that I have on a public server (uses its own database, not the production database). I'll get that set up today.

**Comment by Alys** [comment]
- [x] As shown in the screenshot below, there should be more space to the left of the "Remove" button and less space before the managers' names in that section. There should also be more space to the left of the "Add Manager" button.
[screenshot](https://cloud.githubusercontent.com/assets/1495809/24175784/dd0deb26-0ee3-11e7-85bb-e86f58274ee9.png)

- [x] Also as shown there, the leader can add themself as manager. Is that intentional? On the one hand, it's undesirable because they leader might think they need to and that wastes their time. On the other hand, if the leader is about to transfer leadership to another member, they might want to make themself a manager first. However that would only happen after the group plan had been cancelled, and it should probably then be up to the new leader whether the old one is allowed to be a manager. I'm leaning towards making it not possible for the leader to declare themself a manger too.

- [x] When a task requiring approval is completed, only the group leader gets a notification that an approval is waiting; managers do not get approvals. If that's deliberate, that's fine but can you let us know if it's deliberate so that we can document it? If there's anything else non-obvious that you think should be documented about this, can you edit the top post to give details about the new features?

- [x] When a task awaiting approval is approved by a manager, the notification that approval is needed is not removed from the group leader's notification list. This is going to cause wasted time and annoyance for the leader. Possible solutions include:
- Give all the group plans notifications "x" buttons so that they can be more easily dismissed - if none of the other options are done, I think this will be essential to reduce the complaints. Even if one of the other options is done, this will still be a good idea now.
- Remove the notification when the task is approved by someone else.
- Don't have individual notifications for any tasks awaiting approval, but instead have one notification for each group plan stating that there is at least one task to be approved. As soon as all tasks are approved, that notification could be automatically removed (or at the very least could be dismissable with an "x" button).

- [x] If a member completes a task requiring approval and it's approved by both a manager and the leader, the member receives twice as much XP and GP for it (i.e., both approvals give rewards to the user). They also receive two notifications that it was approved. This can only happen if the second approver had not refreshed the list of approvals before approving, but that could happen in real life when two approvers happened to be working through the list of pending approvals at the same time.

**Comment by Alys** [comment]
Oh, this comment might be irrelevant: "However that would only happen after the group plan had been cancelled". That's from the part where I was saying that the leader can add themself as manager.

- [x] I see now that managers also be added to public guilds and parties. Assuming that's deliberate, what functions can they do?

File: website/client-old/js/components/groupTasks/groupTasksController.js
Comment (Alys) [inline website/client-old/js/components/groupTasks/groupTasksController.js]: Can `sup` be given a more meaningful name? I can't work out from the code what it's for so a longer name might help.

File: website/server/controllers/api-v3/tasks.js
Comment (Alys) [inline website/server/controllers/api-v3/tasks.js]: Can `isNotGroupLeader` be changed to `isGroupLeader`? That would be consistent with `isManager`.

File: website/client-old/js/components/groupTasks/groupTasksController.js
Comment (Alys) [inline website/client-old/js/components/groupTasks/groupTasksController.js]: Can this be changed to be more like `isManager = Boolean(group.managers[user._id]);` for consistency with that code?

**Review by Alys — COMMENTED** [review]
I haven't actually done a code review. These are just a couple of things that I noticed when I was looking through the code to work out how the new features work. :)

File: website/client-old/js/components/groupTasks/groupTasksController.js
Comment (TheHollidayInn) [reply website/client-old/js/components/groupTasks/groupTasksController.js]: Yea... That is an accident lol

File: website/server/controllers/api-v3/groups.js
Comment (paglias) [inline website/server/controllers/api-v3/groups.js]: since these errors are not meant for end users, can we use the new `apiMessages` lib like here https://github.com/HabitRPG/habitica/commit/939712ad1f8e8c8e228907b30844fcd3bb5dae51#diff-b950a7c557b35ea58bc2685532c46170R285 So the i18n strings to load on the client don't grow in size. It should be used for all API messages meant for devs that don't need to be translated

File: website/server/controllers/api-v3/groups.js
Comment (paglias) [inline website/server/controllers/api-v3/groups.js]: selecting `guilds` should be enough in this case

File: website/server/controllers/api-v3/tasks/groups.js
Comment (paglias) [inline website/server/controllers/api-v3/tasks/groups.js]: can we abstract these lines into a function since it's used in 4 places in this file?

File: website/server/models/group.js
Comment (paglias) [inline website/server/models/group.js:108]: this field should be added to the `noSet` list

**Review by paglias — CHANGES_REQUESTED** [review]

File: website/server/controllers/api-v3/groups.js
Comment (Alys) [reply website/server/controllers/api-v3/groups.js]: Parties can have group plans too. Is that relevant here?

Comment (paglias) [reply website/server/controllers/api-v3/groups.js]: maybe... I said guilds because it only checks for them 7 lines below this one

File: website/server/controllers/api-v3/groups.js
Comment (paglias) [inline website/server/controllers/api-v3/groups.js]: group.isMember should be used here https://github.com/HabitRPG/habitica/blob/develop/website/server/models/group.js#L369

**Comment by Alys** [comment]
- [x] When a party leader upgrades their party, they cannot use this new feature to make a party member a Manager. When they try, they see the message "User must be a member" (I checked that the user I was trying to make a manager really was a member).

File: website/server/controllers/api-v3/tasks/groups.js
Comment (TheHollidayInn) [reply website/server/controllers/api-v3/tasks/groups.js]: Yea, we haven't been abstracted our permission checks, but I'm fine getting that started.

File: website/server/libs/apiMessages.js
Comment (paglias) [inline website/server/libs/apiMessages.js:11]: can you remove the strings from the english locale now that they've been added here?

**Review by paglias — CHANGES_REQUESTED** [review]

File: website/server/libs/apiMessages.js
Comment (TheHollidayInn) [reply website/server/libs/apiMessages.js:11]: This local is used in multiple places not just my routes.

Comment (paglias) [reply website/server/libs/apiMessages.js:11]: Ah right

File: website/server/controllers/api-v3/tasks/groups.js
Comment (Alys) [inline website/server/controllers/api-v3/tasks/groups.js]: Can `isNotGroupLeader` be changed to `isGroupLeader` ?

File: website/server/controllers/api-v3/tasks/groups.js
Comment (Alys) [inline website/server/controllers/api-v3/tasks/groups.js]: This `canNotEditTasks` code looks like there's a new feature to stop the group leader doing something they used to be able to. Is that right? If so, can you please write something about that in the PR so that we have it documented? If a behaviour has changed, we'll need to update the wiki and make sure the socialites know about it so that they can answer questions (also Sabe and I need to know so that we can handle emailed questions).

File: website/client-old/js/components/groupTasks/groupTasksController.js
Comment (Alys) [inline website/client-old/js/components/groupTasks/groupTasksController.js]: Can `userIsManger` be just `isManager` for consistency with other code? If they all have the same name, finding all places where the check is made will be easier, if we ever need to do that. (Also in the other places where this appears.)

File: website/common/locales/en/limited.json
Comment (Alys) [inline website/common/locales/en/limited.json]: I'm not sure about this but it looks like `managerIdRequired` should be in apiMessages.json. Althernatively, if it is meant to be seen by the users and translated, we should make the message more user-friendly.

**Review by Alys — CHANGES_REQUESTED** [review]

File: website/server/controllers/api-v3/tasks/groups.js
Comment (TheHollidayInn) [reply website/server/controllers/api-v3/tasks/groups.js]: Nope, this is not a change. If they are a group leader they can edit tasks. In English this says,

If you are not a group leader or a manager or assigning yourself, then you cannot assign this task.

File: website/common/locales/en/limited.json
Comment (TheHollidayInn) [reply website/common/locales/en/limited.json]: Paglias said to put it here in a comment above.

Comment (TheHollidayInn) [reply website/common/locales/en/limited.json]: It is not meant to be translated.

**Comment by TheHollidayInn** [comment]
" I see now that managers also be added to public guilds and parties. Assuming that's deliberate, what functions can they do?"

@Alys Parties can have group plans. And managers will be able to do other things as well in the future. I can hide this options on the UI if they are not a group plan, I guess.

File: website/server/controllers/api-v3/tasks/groups.js
Comment (Alys) [reply website/server/controllers/api-v3/tasks/groups.js]: Ah, sorry, that makes sense. I was confused by the mix of editing and assigning terminology here.

Is this line a concern here?
`let userIsAssigningToSelf = Boolean(assignedUserId && user._id === assignedUserId);`
Since the canNotEditTasks() function is used to control editing as well as assigning, will that line allow an ordinary member to edit the task if they are assigning it to themself? Even if the code is safe now, is it possible it could introduce that bug in future if the function is used in ways that it's not being used now? It feels odd that a function labelled as being for controlling editing is also being used to allow other permissions to users who should never be allowed to edit.

Would it be safer to have two functions, one to control editing and the other to control assigning? The function to control assigning could use canNotEditTasks() for most of its checks. At the least, I think this function needs to be renamed or commented to accurately describe what it's for.

**Comment by Alys** [comment]
Yes, please hide the options for guilds and parties that don't have group plans. If they're visible, we'll get a lot of questions about them and complaints that they don't do anything.

File: website/common/locales/en/limited.json
Comment (Alys) [reply website/common/locales/en/limited.json]: It should go in the apiMessages file. Some translators would try to convert `req.body.managerId` to their own language.

File: website/server/controllers/api-v3/tasks/groups.js
Comment (TheHollidayInn) [reply website/server/controllers/api-v3/tasks/groups.js]: Yea, that makes sense. The abstraction was completely correct. I'll split the functions.

**Comment by TheHollidayInn** [comment]
I just pushed an update that has the following additions (mangers include leader):
1. All managers will get notifications
2. Only one manger will be able to approve a task
3. All managers will have the notification remove if anyone approves
4. Parties can now use managers correctly

File: website/common/locales/en/limited.json
Comment (paglias) [reply website/common/locales/en/limited.json]: Yeah it should go into `apiMessages`, I had only looked at the other file previously

File: website/client-old/js/components/groupTasks/groupTasksController.js
Comment (TheHollidayInn) [reply website/client-old/js/components/groupTasks/groupTasksController.js]: Not exactly user what you mean. I'm fine changing the variable, but this function is abstracted so the variable definition is only here.

File: website/common/locales/en/limited.json
Comment (paglias) [inline website/common/locales/en/limited.json]: this should be removed now that it's been moved to apiMessages

File: website/common/locales/en/limited.json
Comment (paglias) [inline website/common/locales/en/limited.json]: all these strings don't seem to be used anywhere in the PR

**Review by paglias — CHANGES_REQUESTED** [review]

File: website/common/locales/en/limited.json
Comment (TheHollidayInn) [reply website/common/locales/en/limited.json]: Not sure what happened here. These appeared when I merged in Devleop.

File: website/common/locales/en/limited.json
Comment (TheHollidayInn) [reply website/common/locales/en/limited.json]: I have no idea why these strings are here. I must have put them here by accident I guess :P

File: website/common/locales/en/limited.json
Comment (paglias) [reply website/common/locales/en/limited.json]: I think these should be removed too, probably some glitch with develop

**Comment by Alys** [comment]
Once the code changes are done, I'd like to test again locally before this is merged. @TheHollidayInn Should I do that now or are there more changes to come?

**Comment by TheHollidayInn** [comment]
Should be ready for more testing! Thanks 👍

File: website/server/controllers/api-v3/tasks/groups.js
Comment (paglias) [inline website/server/controllers/api-v3/tasks/groups.js]: there isn't a better way to identify the notification? Relaying on the translated text doesn't seem optimal and would fail if the user changes its language

File: website/server/controllers/api-v3/tasks/groups.js
Comment (paglias) [inline website/server/controllers/api-v3/tasks/groups.js]: can we just fetch notifications?

File: website/server/controllers/api-v3/tasks/groups.js
Comment (paglias) [reply website/server/controllers/api-v3/tasks/groups.js]: Actually this won't work as soon as one manager has a language different than the others

File: website/server/controllers/api-v3/tasks/groups.js
Comment (paglias) [inline website/server/controllers/api-v3/tasks/groups.js:21]: this function (just statements outside of ay function) seems also used in  website/server/controllers/api-v3/tasks.js, can we abstract it somewhere and add some unit tests?

File: website/server/controllers/api-v3/tasks.js
Comment (paglias) [inline website/server/controllers/api-v3/tasks.js]: let's just load preferences and notifications?

**Review by paglias — CHANGES_REQUESTED** [review]

File: website/server/controllers/api-v3/tasks/groups.js
Comment (TheHollidayInn) [reply website/server/controllers/api-v3/tasks/groups.js]: Yep, I added a comment about that.

File: website/server/controllers/api-v3/tasks/groups.js
Comment (TheHollidayInn) [reply website/server/controllers/api-v3/tasks/groups.js:21]: If we start abstracting permissions for routes, I'd rather have a full separate PR on creating "policy" services in a unified way.

Comment (paglias) [reply website/server/controllers/api-v3/tasks/groups.js:21]: In general I'd say okay but this PR is adding the same code just written differently in two different places. It won't require much effor to create a function like `user.isSubscribed()` and avoid confusion later

File: website/server/controllers/api-v3/tasks/groups.js
Comment (paglias) [reply website/server/controllers/api-v3/tasks/groups.js]: so what happens if two managers have different languages? the notification cannot be removed?

Comment (TheHollidayInn) [reply website/server/controllers/api-v3/tasks/groups.js]: Oh, wrong section. I see what you mean. Let me add the task_id instead.

File: website/server/controllers/api-v3/tasks/groups.js
Comment (TheHollidayInn) [reply website/server/controllers/api-v3/tasks/groups.js:21]: IT would be easy to abstract, just incorrect. user.isLeader, user.isManager make sense, but user.canNotEditTasks doesn't really make sense for the owning object to be the approver. 

So, you could add the function to the group and have group.userCanEditTask, but then we see that this code is duplicated as well: https://github.com/HabitRPG/habitica/blob/develop/website/server/controllers/api-v3/tasks.js#L1192 . So, why are we abstracting one and not the other? 

Then, we run into a SOLID design issue on the group model (which we already have). But we can see that a better (first step) would be to create a TaskPolicy class to begin handling these permissions. Which would be out of scope of this PR.

So, sure, it is easy, but it would be support bad code that I have already brought up as an issue.

File: test/api/v3/integration/tasks/groups/POST-group_tasks_id_score_direction.test.js
Comment (Alys) [inline test/api/v3/integration/tasks/groups/POST-group_tasks_id_score_direction.test.js]: Typo: notificatoins > notifications

File: website/server/libs/apiMessages.js
Comment (Alys) [inline website/server/libs/apiMessages.js]: Should `canOnlyApproveTaskOnce` be translatable? It sounds like a message the group leader will see.

The text would be better as "This task has already been approved." 
("You can only approve a task once" will confuse a user who thinks it's saying that they themself have tried to approve it twice.)

File: website/views/options/social/group.jade
Comment (Alys) [inline website/views/options/social/group.jade]: This blank line has whitespace in it (its a really minor, picky point but I think we try to avoid that, and git shows it as a red mark). Line 78 at the end of this block also has whitespace.

**Review by Alys — CHANGES_REQUESTED** [review]

**Comment by Alys** [comment]
This works:
	"All managers will have the notification remove if anyone approves"
except for two things:
- [x] The notification disappears only when you reload, not when you sync. Can it be made to disappear on sync? Reloading isn't a common action at the moment.
- [x] The notification doesn't disappear if the manager has been demoted before approval was done. Steps to reproduce:
   1. The leader makes user Bob a manager.
   2. Someone completes a task requiring approval.
   3. All managers, including Bob, get a notification that approval is needed.
   4. Before the task is approved, the leader removes Bob's management status. Bob still has the notification.
   6. The task is approved.
   7. The notification is removed from all **current** managers. Bob still has the notification.
   - This could be fixed by removing the notification at step 4 (probably best) or at step 7. It's an edge case though, so if both of those are difficult to achieve, we could ignore it.

I'd really like it if "x" buttons could be added to the notifications in this PR, if it's not very difficult. In groups of peers, potentially all group members could be managers, and even with the notifications disappearing automatically when approvals are done, the notifications are still going to seem spammy if they can't also be deleted manually.

----

I think to make the manager feature easy for group members to follow, the group membership list needs to state who is the leader and who are the managers. For example, something like this:

image

The "leader" indicator should appear on all parties and guilds, not just those that use group plans or that have managers.

I suggest that the "manager" indicator should also be coded so that it could work in all groups, not just those that have group plans (even though managers currently can't be assigned in non-group-plan groups). That way the indicators will work immediately once managers are enabled for all groups.

**Comment by Alys** [comment]
- [x] The leader and manager indicators in the group membership list would be better at the right of the HP indicator, otherwise mischievous players can make themselves look like a leader or manager by changing their name, and I am certain we have some players who would do that. For example in the image below, Canasta is not the leader. It would probably be best if the leader/manager indicators had a different appearance as well, for example, normal text rather than being part of the link.

image


@TheHollidayInn Are you still working through the previous suggestions or is this ready to be tested again?

**Review by paglias — APPROVED** [review]

**Comment by Alys** [comment]
Ah heck I'm so sorry. I tried to push a commit with some minor typo fixes and one text change and I've screwed it up by pushing recent commits from develop. :( 

@TheHollidayInn I think if you force push from your local group-managers branch to your remote group-managers branch it will clean everything up.

The commit I was trying to push was this one: https://github.com/HabitRPG/habitica/pull/8591/commits/595441ab59c10822088de35c43d5439d41e217e1

I've tested and everything else looks good except for a small problem with notifications:

When a completed task is approved, all managers and leader are supposed to be able to sync to remove the notification about the task needing approval. I've found that doesn't work all the time though. I'm not sure this is consistent but it seems that usually when there's only one notification about a pending task approval, then the notification is often not removed by sync. If there's two or more notifications, each one is removed by a sync.
If you click a notification that should have been removed, you get a "notification" not found message and then the notification disappears, so this isn't a critical error. If it's complex to fix, I don't think it needs to be fixed for this PR to be merged, but maybe an issue should be logged for it. It might be part of a wider problem with our notification system - that might be what one of the new comments in the code is talking about. :)

**Pull request merged**
